webhook/resources/backuptarget/validator.go

```diff
@@ -0,0 +1,276 @@
+package backuptarget
+
+import (
+	"fmt"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	"github.com/longhorn/longhorn-manager/types"
+	"github.com/longhorn/longhorn-manager/util"
+	"github.com/longhorn/longhorn-manager/webhook/admission"
+
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	werror "github.com/longhorn/longhorn-manager/webhook/error"
+)
+
+type backupTargetURLType string
+
+const (
+	backupTargetURLTypeBucket  = backupTargetURLType("bucket")
+	backupTargetURLTypeOptions = backupTargetURLType("options")
+)
+
+type backupTargetValidator struct {
+	admission.DefaultValidator
+	ds *datastore.DataStore
+}
+
+func NewValidator(ds *datastore.DataStore) admission.Validator {
+	return &backupTargetValidator{ds: ds}
+}
+
+func (b *backupTargetValidator) Resource() admission.Resource {
+	return admission.Resource{
+		Name:       "backuptargets",
+		Scope:      admissionregv1.NamespacedScope,
+		APIGroup:   longhorn.SchemeGroupVersion.Group,
+		APIVersion: longhorn.SchemeGroupVersion.Version,
+		ObjectType: &longhorn.BackupTarget{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Create,
+			admissionregv1.Update,
+		},
+	}
+}
+
+func (b *backupTargetValidator) Create(request *admission.Request, newObj runtime.Object) error {
+	backupTarget := newObj.(*longhorn.BackupTarget)
+
+	if !util.ValidateName(backupTarget.Name) {
+		return werror.NewInvalidError(fmt.Sprintf("invalid name %v", backupTarget.Name), "")
+	}
+
+	if err := b.validateBackupTargetURL(backupTarget.Spec.BackupTargetURL); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	if err := b.validateCredentialSecret(backupTarget.Spec.CredentialSecret); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	if err := b.handleAWSIAMRoleAnnotation(backupTarget, nil); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	return nil
+}
+
+func (b *backupTargetValidator) Update(request *admission.Request, oldObj runtime.Object, newObj runtime.Object) error {
```

Comment:
We should 
1. check if the current `BackupTarget` is default, it it is, we should not allow user to change the url to empty
2. change the `BackupTarget.Spec.Default` from `true` to `false`

Reply:
oh that's right, user can still make the url empty and the backupTarget will become available=false

Reply:
Added this logic at the validator.